executor/include/executor/property/blocking.hpp

```diff
@@ -13,48 +13,46 @@
 
 namespace executor {
 
-struct blocking_t {
- public:
+// Part of Proposal P0443R13: 2.2.12.1
+struct blocking_t : basic_executor_property<blocking_t, false, false> {
+  friend constexpr bool operator==(const blocking_t& a, const blocking_t& b) {
+    return a.value_ == b.value_;
+  }
+
+  friend constexpr bool operator!=(const blocking_t& a, const blocking_t& b) {
+    return !(a == b);
+  }
+
+  constexpr blocking_t() : value_{0} {};
+
   struct always_t : basic_executor_property<always_t, true, true> {
-    template <class Executor>
-    friend Executor require(Executor& ex, const always_t& t) {
-      return ex.require(t);
-    }
-    template <class Executor>
-    friend bool query(const Executor& ex, const always_t& t) {
-      return std::is_same<always_t, decltype(ex.interface)>();
-    }
+    static constexpr always_t value() { return {}; }
   };
 
-  const always_t always;
+  static constexpr always_t always{};
+  constexpr blocking_t(const always_t&) : value_{1} {};
 
   struct never_t : basic_executor_property<never_t, true, true> {
-    template <class Executor>
-    friend Executor require(Executor& ex, const never_t& t) {
-      return ex.require(t);
-    }
-    template <class Executor>
-    friend bool query(const Executor& ex, const never_t& t) {
-      return std::is_same<never_t, decltype(ex.interface)>();
-    }
+    static constexpr never_t value() { return {}; }
   };
 
-  const never_t never;
+  static constexpr never_t never{};
+  constexpr blocking_t(const never_t&) : value_{2} {};
 
   struct possibly_t : basic_executor_property<possibly_t, true, true> {
-    template <class Executor>
-    friend Executor require(Executor& ex, const possibly_t& t) {
-      return ex.require(t);
-    }
-    template <class Executor>
-    friend bool query(const Executor& ex, const possibly_t& t) {
-      return std::is_same<never_t, decltype(ex.interface)>();
-    }
+    static constexpr possibly_t value() { return {}; }
   };
 
-  const possibly_t possibly;
+  static constexpr possibly_t possibly{};
+  constexpr blocking_t(const possibly_t&) : value_{3} {};
+
+  int value_;
```

Comment:
There are some references to a sort of numbering in the proposal, in `P0443R13: 2.2.12.1`. I had picked this up from https://github.com/executors/executors-impl.
I haven't quite understood all the requirements so in the current state only some requirements of `Property` are satisfied.

> The value of the expression S::Ni::static_query_v<Executor>, for all 1 < i <= N, is
> 
> Executor::query(S::Ni()), if that expression is a well-formed constant expression;
> otherwise ill-formed.
> The value of the expression S::static_query_v<Executor> is
> 
> Executor::query(S()), if that expression is a well-formed constant expression;
> otherwise, ill-formed if declval<Executor>().query(S()) is well-formed;
> otherwise, S::Ni::static_query_v<Executor> for the least i <= N for which this expression is a well-formed constant > > > expression;
> otherwise ill-formed.

Reply:
That section states that there are generic N mutually exclusive behavioral properties. No that there's a notion of explicit order i.e. `N1 < N2 < ... < NN`

> There are some references to a sort of numbering in the proposal, in `P0443R13: 2.2.12.1`. I had picked this up from https://github.com/executors/executors-impl.

Please provide a url to the actual lines of code where the ordering implementation takes place.

Reply:
For cudex: https://github.com/jaredhoberock/cudex/blob/b98bcdd558a9045215b12608727d74718f3867f6/cudex/property/bulk_guarantee.hpp#L54-L57

For executor-impl: https://github.com/executors/executors-impl/blob/58a2b06dfe26c6d3d62bc0926527c1c9eb7fdf0b/include/bits/execution/blocking.h#L17-L19